gio-subclass/src/action.rs

```diff
@@ -0,0 +1,202 @@
+// This file was generated by gir (https://github.com/gtk-rs/gir @ ec3c0fa+)
+// from gir-files (https://github.com/gtk-rs/gir-files @ 2b56823)
+// DO NOT EDIT
+
+#[allow(unused_imports)]
+use libc::{c_int, c_char, c_uchar, c_float, c_uint, c_double,
+    c_short, c_ushort, c_long, c_ulong,
+    c_void, size_t, ssize_t, intptr_t, uintptr_t, time_t, FILE};
+
+#[allow(unused_imports)]
+use glib::{gboolean, gconstpointer, gpointer, GType};
+
+use gobject_subclass::anyimpl::*;
+use gobject_subclass::object::*;
+
+#[cfg(any(feature = "v2_38", feature = "dox"))]
+use Error;
+use ffi;
+use glib;
+use glib::object::Downcast;
+use glib::object::IsA;
+use glib::signal::SignalHandlerId;
+use glib::signal::connect;
+use glib::translate::*;
+use glib_ffi;
+use gobject_ffi;
+use std::boxed::Box as Box_;
+use std::mem;
+use std::mem::transmute;
+use std::ptr;
+
+
+pub trait ActionImpl: AnyImpl + 'static {
+
+    fn activate(&self, action: &Action, parameter: Option<&glib::Variant>);
+
+    fn change_state(&self, action: &Action, value: &glib::Variant);
+
+    fn get_enabled(&self, action: &Action) -> bool;
+
+    fn get_name(&self, action: &Action) -> Option<String>;
+
+    fn get_parameter_type(&self, action: &Action) -> Option<glib::VariantType>;
+
+    fn get_state(&self, action: &Action) -> Option<glib::Variant>;
+
+    fn get_state_hint(&self, action: &Action) -> Option<glib::Variant>;
+
+    fn get_state_type(&self, action: &Action) -> Option<glib::VariantType>;
+
+}
+
+any_impl!(ActionImpl);
+
+pub trait ActionImplStatic<T: ObjectType>: 'static {
+    fn get_impl<'a>(&self, imp: &'a T::ImplType) -> &'a ActionImpl;
+}
+
+struct ActionStatic<T: ObjectType>{
+    imp_static: *const ActionImplStatic<T>
+}
+
+unsafe extern "C" fn action_activate<T: ObjectType>
+(ptr: *mut GAction, parameter: *mut glib::GVariant)
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let klass = &**(ptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(ffi::action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.activate(&wrap, &from_glib_none(action), /*Unknown conversion*/parameter)
```

Comment:
This could be `&from_glib_borrow(action)` (and IIRC would be for signals). Not sure why it falls apart on the `GVariant`, in the function just below it works :)

Reply:
Generally this should do `&from_glib_borrow` for all shared/boxed/object types like it would in signal trampolines. There's probably a code path that has to be enabled for your case too here

Reply:
See e.g. https://github.com/sdroege/gstreamer-rs/blob/85ca3b94a34d0236314aac900d5926c81ba73d99/gstreamer/src/auto/element.rs#L535-L540

Reply:
Actually, all of the 'action' conversions here are incorrect. That's what 'wrap' is

Reply:
Oh! Indeed